metadata-ingestion/src/datahub/cli/docker_check.py

```diff
@@ -1,122 +1,203 @@
-    # need to make sure that we only yield once.
 
-    docker_cli = None
+class DockerNotRunningError(Exception):
+    SHOW_STACK_TRACE = False
+
+
+class DockerLowMemoryError(Exception):
+    SHOW_STACK_TRACE = False
+
+
+class DockerComposeVersionError(Exception):
+    SHOW_STACK_TRACE = False
+
+
+class QuickstartError(Exception, ExceptionWithProps):
+    SHOW_STACK_TRACE = False
+
+    def __init__(self, message: str, container_statuses: Dict[str, Any]):
+        super().__init__(message)
+        self.container_statuses = container_statuses
+
+    def get_telemetry_props(self) -> Dict[str, Any]:
+        return self.container_statuses
+
+
+@contextmanager
+def get_docker_client() -> Iterator[docker.DockerClient]:
+    # Get a reference to the Docker client.
+    client = None
     try:
-        docker_cli = docker.from_env()
+        client = docker.from_env()
     except docker.errors.DockerException as error:
         try:
             # Docker Desktop 4.13.0 broke the docker.sock symlink.
             # See https://github.com/docker/docker-py/issues/3059.
             maybe_sock_path = os.path.expanduser("~/.docker/run/docker.sock")
             if os.path.exists(maybe_sock_path):
-                docker_cli = docker.DockerClient(base_url=f"unix://{maybe_sock_path}")
+                client = docker.DockerClient(base_url=f"unix://{maybe_sock_path}")
             else:
-                yield None, error
+                raise error
         except docker.errors.DockerException as error:
-            yield None, error
+            raise DockerNotRunningError(
+                "Docker doesn't seem to be running. Did you start it?"
+            ) from error
+    assert client
 
-    if docker_cli is not None:
-        try:
-            docker_cli.ping()
-        except docker.errors.DockerException as error:
-            yield None, error
-        else:
-            try:
-                yield docker_cli, None
-            finally:
-                docker_cli.close()
+    # Make sure that we can talk to Docker.
+    try:
+        client.ping()
+    except docker.errors.DockerException as error:
+        raise DockerNotRunningError(
+            "Unable to talk to Docker. Did you start it?"
+        ) from error
+
+    # Yield the client and make sure to close it.
+    try:
+        yield client
+    finally:
+        client.close()
 
 
 def memory_in_gb(mem_bytes: int) -> float:
     return mem_bytes / (1024 * 1024 * 1000)
 
 
-def check_local_docker_containers(preflight_only: bool = False) -> List[str]:
-    issues: List[str] = []
-    with get_client_with_error() as (client, error):
-        if error:
-            issues.append("Docker doesn't seem to be running. Did you start it?")
-            return issues
+def run_quickstart_preflight_checks(client: docker.DockerClient) -> None:
+    # Check total memory.
+    # TODO: add option to skip this check.
+    total_mem_configured = int(client.info()["MemTotal"])
```

Comment:
Confirming that "MemTotal" will always exist inside the info dict. If not, we should use a default like... 0

Reply:
Also, should we be casting to int? Above our min is 3.8 right?

Reply:
Nvm i see it _is_ an int - disregard